### ISSUE TITLE
vinyl: replace `struct vy_stmt` with `struct tuple`

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#1572]
Our strategic goal is to replace in-memory level in Vynil with MemTX data structures. 
@alyapunov integrated matras and bps-tree for memtable, I removed phia_document and switched format to MessagePack. What is about replacing `struct phia_tuple` (former `struct svv`) with `struct tuple`?

```
struct vinyl_tuple {
        uint64_t lsn;
        uint32_t size;
        uint16_t refs; /* atomic */
        uint8_t  flags;
        char data[0];
};

struct PACKED tuple
        /** snapshot generation version */
        uint32_t version;
        /** reference counter */
        uint16_t refs;
        /** format identifier */
        uint16_t format_id;
        /** length of the variable part of the tuple */
        uint32_t bsize;
        char data[0];
};
```

There are three problems:
1. tuple_alloc()/tuple_free() is not thread safe
   Possible solutions: 
   a) add virtual malloc-based tuple_alloc()/tuple_free() to tuple_format
   b) Add mutex - **affects existing users**.
   c) Implement delayed removal of svvindex from TX thread
2. tuple_ref()/tuple_unref() is not thread safe
   Possible solutions: 
   a) add atomics - **affects existing users**.
   b) Implement delayed removal of svvindex from TX thread
3. phia tuple has `int64_ lsn`, memtx tuple has `int32_t version`
   Possible solution: change `int32_t version` to `int64_t version` in `struct tuple` - **affects existing users**.

I dislike solution with multithread alloc()/free() ref()/unref(). We need to discuss problems above to move on.

**Comment by rtsisyk** [comment]
Postponed.

**Comment by rtsisyk** [comment]
Stage 1: convert public vinyl API to accept `struct tuple` / `(const char *key,  uint32_t part_count)` and return `struct tuple` instead of `struct vinyl_tuple`, hide `struct vinyl_tuple` inside `vinyl.c`.

**Comment by rtsisyk** [comment]
Stage 1 is finished, `struct tuple` conversion is postponed.

**Comment by rtsisyk** [comment]
Blocked by #1685

**Comment by rtsisyk** [comment]
vy_tuple now used only from TX thread.

**Comment by rtsisyk** [comment]
## Stage 2

Actually, vy_tuple is an vinyl operation - REPLACE, DELETE, GET or UPSERT. I propose to supercede tuples with stamemts (stmt):

- [ ] Rename `struct vy_tuple` to `struct vy_stmt`
- [ ] Replace `tuple->flags` with `uint32_t stmt->type`
- [ ] Refactor all tuples methods to vy_stmt_new_TYPE()/vy_stmt_TYPE_action() style:
   - * vy_tuple_from_data() => vy_stmt_new_replace()
   - * vy_tuple_from_data_ex() => vy_stmt_new_upsert()
   - * vy_tuple_from_key_key() => vy_stmt_new_get()/vy_stmt_new_delete()
   - * vy_tuple_key_part() => vy_stmt_key_part()
   - * vy_tuple_data() => vy_stmt_replace_tuple()
   - * vy_tuple_data_ex() => vu_stmt_upsert_tuple() + vy_stmt_upsert_ops()
   - * vy_tuple_compare() -> vy_stmt_compare()

## Stage 3

- [ ] Split vy_tuple_compare() to vy_stmt_compare_tuple(), vy_stmt_compare_key(), vy_stmt_compare_tuple_with_key().  Detect combination in runtime using stmt->type designator.
- [ ] Split vy_tuple_key_part() to two parts (GET + DELETE, REPLACE + UPSERT)
- [ ] Use `const char *key, uint32_t part_count` format for GET and DELETE
- [ ] Use `struct tuple` format for tuple in REPLACE and UPSERT, raw msgpack for UPSERT ops

**Comment by rtsisyk** [comment]
Stage 3 is over.

```
struct PACKED tuple
{
        /*
         * sic: the header of the tuple is used
         * to store a free list pointer in smfree_delayed.
         * Please don't change it without understanding
         * how smfree_delayed and snapshotting COW works.
         */
        /** snapshot generation version */
        uint32_t version;
        /** reference counter */
        uint16_t refs;
        /** format identifier */
        uint16_t format_id;
        /** length of the variable part of the tuple */
        uint32_t bsize;
        /**
         * Fields can have variable length, and thus are packed
         * into a contiguous byte array. Each field is prefixed
         * with MessagePack packed field length.
         */
        char data[0];
};
```

```
struct vy_stmt {
        int64_t  lsn;
        uint32_t size;
        uint16_t refs; /* atomic */
        uint8_t  type; /* IPROTO_SELECT/REPLACE/UPSERT/DELETE */
        char data[0];
};
```

We need 4 more bytes for lsn and one more byte for type to replace `struct vy_stmt` with `struct tuple`.

**Comment by rtsisyk** [comment]
Actually format_id is not needed for vinyl, all tuples are always in the same format.

**Comment by rtsisyk** [comment]
Discussed today => postponed

**Comment by rtsisyk** [comment]
A spec based on @kostja proposal:

```
struct PACKED tuple {
	/** length of the variable part of the tuple */
	uint32_t bsize;
	/** reference counter */
	uint16_t refs;
	/** format identifier */
	uint16_t format_id;
	/** offset of msgpack relative to ->data */
	uint8_t offset;
	/**
         * Format-specific meta and msgpack data
         * @sa tuple_meta_memtx
         * @sa tuple_meta_vinyl
         */
	char data[0];
};

struct PACKED tuple_meta_memtx {
	/** snapshot generation version */
	uint32_t version;
	/** offsets table - see tuple_format **/
	uint32_t offsets[0];
};

struct PACKED vy_tuple_vinyl {
	/** lsn for MVCC */
	int64_t lsn;
	/** statement type - IPROTO_SELECT/REPLACE/UPSERT/DELETE */
	int8_t type;
	/** offsets table - see tuple_format **/
	uint32_t offsets[0];
};
```

struct vy_stmt will be completely removed. A designator of engine (memtx vs vinyl) should be stored in `struct tuple_format`. 

Note that SELECT and DELETE statements doesn't require offsets. Please register a special format_id to check by assertions.

**Comment by rtsisyk** [comment]
- [ ] Introduce tuple_data() - a function to get MessagePack array from `struct vy_tuple`:
```
const char *
struct tuple_data(const char *tuple);
```
- [ ] Remove direct access to `tuple->data`, rename `tuple->data` to `tuple->raw`
- [ ] Add `tuple->offset` member, split `tuple->version` into `struct tuple_meta_memx`
- [ ] Move offsets to struct tuple_meta_memtx, remove hacks with negative offsets
- [ ] Implement support for `struct tuple_meta_vinyl` in vy_stmt_XXX wrappers
- [ ] Discard `vy_stmt` (probably huge patch)
- [ ] Remove tuple -> tuple -> tuple conversion in vinyl public API (PROFIT!)

See fix-gh-1386-tuple-restructure-v7 branch from former 1.8 for inspiration for the first points.
